src/Algebra/Ordered/Structures.agda

```diff
@@ -0,0 +1,205 @@
+
+    isPreorder : IsPreorder _≈_ _∼_
+    isPreorder = record
+      { isEquivalence = isEquivalence
+      ; reflexive     = reflexive
+      ; trans         = trans
+      }
+
+    open IsPreorder isPreorder public hiding (isEquivalence; reflexive; trans)
+
+  -- Preordered monoids (promonoids)
+
+  record IsPromonoid (_∙_ : Op₂ A) (ε : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isProsemigroup : IsProsemigroup _∙_
+      identity       : Identity _≈_ ε _∙_
+
+    open IsProsemigroup isProsemigroup public
+
+    isMonoid : IsMonoid _≈_ _∙_ ε
+    isMonoid = record
+      { isSemigroup = isSemigroup
+      ; identity    = identity
+      }
+
+    open IsMonoid isMonoid public using (identityˡ; identityʳ)
+
+  record IsCommutativePromonoid (_∙_ : Op₂ A) (ε : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isPromonoid : IsPromonoid _∙_ ε
+      comm        : Commutative _≈_ _∙_
+
+    open IsPromonoid isPromonoid public
+
+    isCommutativeMonoid : IsCommutativeMonoid _≈_ _∙_ ε
+    isCommutativeMonoid = record
+      { isMonoid        = isMonoid
+      ; comm            = comm
+      }
+
+  -- Preordered semirings (prosemirings)
+
+  record IsProsemiring (_+_ _*_ : Op₂ A) (0# 1# : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      +-isCommutativePromonoid : IsCommutativePromonoid _+_ 0#
+      *-isPromonoid            : IsPromonoid _*_ 1#
+      distrib                  : _DistributesOver_ _≈_ _*_ _+_
+      zero                     : Zero _≈_ 0# _*_
+
+    distribˡ : _DistributesOverˡ_ _≈_ _*_ _+_
+    distribˡ = proj₁ distrib
+
+    distribʳ : _DistributesOverʳ_ _≈_ _*_ _+_
+    distribʳ = proj₂ distrib
+
+    zeroˡ : LeftZero _≈_ 0# _*_
+    zeroˡ = proj₁ zero
+
+    zeroʳ : RightZero _≈_ 0# _*_
+    zeroʳ = proj₂ zero
+
+    open IsCommutativePromonoid +-isCommutativePromonoid public
+      renaming
+      ( assoc       to +-assoc
+      ; ∙-cong      to +-cong
+      ; monotonic   to +-monotonic
+      ; identity    to +-identity
+      ; identityˡ   to +-identityˡ
+      ; identityʳ   to +-identityʳ
+      ; comm        to +-comm
+      ; isMagma     to +-isMagma
+      ; isSemigroup to +-isSemigroup
+      ; isMonoid    to +-isMonoid
+      )
+
+    open IsPromonoid *-isPromonoid public
+      using ()
+      renaming
+      ( assoc       to *-assoc
+      ; ∙-cong      to *-cong
+      ; monotonic   to *-monotonic
+      ; identity    to *-identity
+      ; identityˡ   to *-identityˡ
+      ; identityʳ   to *-identityʳ
+      ; isMagma     to *-isMagma
+      ; isSemigroup to *-isSemigroup
+      ; isMonoid    to *-isMonoid
+      )
+
+------------------------------------------------------------------------
+-- Partially ordered structures
+
+module _ {a ℓ₁ ℓ₂} {A : Set a}
+  (_≈_ : Rel A ℓ₁)  -- The underlying equality.
+  (_≤_ : Rel A ℓ₂)  -- The partial.
+  where
+
+  -- Partially ordered semigroups (posemigroups)
+
+  record IsPosemigroup (_∙_ : Op₂ A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isPartialOrder : IsPartialOrder _≈_ _≤_
```

Comment:
Hmm the lack of symmetry between this and `IsProsemigroup` is a bit off-putting. The latter uses the algebraic structure `IsSemigroup` as the base structure, whereas this uses the ordered structure `IsPartialOrder` as the base structure. Is there a good reason for this? If not I think it'd be better for them to be consistent!

Reply:
Yea, I see what you mean. The reason is that you get some structure for free once you assume anti-symmetry as well. Notably, you get `∙-cong` for free or really `cong` for any operation that is also monotone (as you noted yourself). There's a tradeoff here between adding redundant fields/assumptions to the records vs. keeping the record structure "regular". I personally prefer not to add redundant fields unless there are good technical reasons to do so (e.g. because it makes some proofs simpler). But I don't feel strongly about this, so I'm happy to discuss alternatives.

Reply:
Yup, I see your point and this is exactly the reason we have `Algebra.Structures.Biased` which provide alternate, "easier" ways of constructing structures when you can derive some of the axioms from the others. So I'd recommend that we preserve the symmetry here, and then add this variant to `Algebra.Ordered.Structures.Biased`? (which is also re-exported from `Algebra.Ordered` for easy use).

Reply:
Alright. Two questions:
 1. Which structure do you prefer: using the algebraic structure as a base and adding what's missing from the orders, or vice-versa? (We could also add both at once, but then `isEquivalence` would be duplicated, which I really don't think is a good idea.)
 2. Is there maybe a better name than `Biased` for this? I don't really see in which sense either version is biased. But I also don't have any good alternative suggestions, TBH...

Reply:
Upon reflection, the answer to 1. is probably to go with preorders as the base since we're going to mirror the whole algebra hierarchy (starting with magmas) and build them up anyway. For the po* structures, I'll similarly start with partial orders.

Reply:
OK, I completely reworked the definition of the various preordered structures using the preorder as a base, adding promagmas (and pomagmas), etc. Surprisingly, it looks like no biased versions are needed after all. 🤷

Reply:
You're going to hate me for this (sorry!), but wouldn't it be much cleaner to have the algebraic structures as the base and then add the ordering on top? The algebraic hierarchy is a mess, and the re-exports/re-naming as you've probably found are incredibly painful. Using the orders as the base structures means we have to replicate it in all its gory detail. Whereas using the algebra as the base structures means that we would get it for "free". The preorder/partial order hierarchy is pretty simple, so re-exporting everything should be much easier.

Reply:
Oh no! 😰

OK, let's think this through before I embark on another refactoring spree. 😅 

I actually don't think using the algebraic structures as a base is such a good idea. For the pre-ordered structures, it would just mean creating a number of records that would simply contain the base algebra plus the axioms of a preorder (using `IsPreorder` directly is not a good idea because it would duplicate the `IsEquivalence` filed which you really don't want, there shouldn't be two versions of that in the same record). But for the partially ordered version, this would mean a redundant `cong` field, which is not a deal breaker, but then we *would* need a separate collection of `Biased` modules that would just replicate what is in the PR now. I think that would actually decrease maintainability overall, no?

Again, I don't have strong feelings about this. If you think it's cleaner the other way round, @MatthewDaggitt, I'll put in the extra effort, but I'd rather we agreed first that's really what we want.

Reply:
> using IsPreorder directly is not a good idea because it would duplicate the IsEquivalence field which you really don't want, there shouldn't be two versions of that in the same record

Agreed!

Hmm so I just wrote 3 paragraphs about why it was better with the algebra hierarchy as a base, but deleted it as I can't distinguish in my head whether I think that because the current algebra hierarchy is so hopelessly broken or whether its actually better that I way. In the end, I've decided at this moment I also don't have the time to work through all the implications (sorry future me!). Let's just keep it the way you've got it now :+1:

Reply:
Yea, I can totally see that there are good arguments for and against here. It's hard to know which will work out better in the end.

Out of curiosity: when you say the algebra hierarchy is broken, do you think there would be a reasonably clean fix, or is it just an impossible problem to design hierarchies like this in Agda? I mean, it does feel like we're just implementing OO/multiple inheritance manually here, but maybe my judgment is clouded from having worked in OO languages too much in the past. In any case, I'm curious to hear if you have any ideas on how things could be improved in general. Is there some discussion around this topic somewhere? E.g. in an issue?

Reply:
> do you think there would be a reasonably clean fix, or is it just an impossible problem to design hierarchies like this in Agda?

I think the consensus is that it's the latter. There's no good way to define diamonds compositionally without biasing them in some regards that impacts usability. I hear that Arend's implementation of record types is supposed to have gone a long way to fixing the problem.

In no particular order here are a selection of the issues discussing problems:
- https://github.com/agda/agda-stdlib/issues/239
- https://github.com/agda/agda-stdlib/issues/455
- https://github.com/agda/agda-stdlib/issues/876
- https://github.com/agda/agda-stdlib/issues/1238
- https://github.com/agda/agda-stdlib/issues/1617

Reply:
Thanks for these pointers. I'll bookmark #1238, which contains a lot of references to relevant related work. (Thanks @JacquesCarette!) If it's not already the case, it might be nice to summarize these discussions in some sort of README.

Reply:
No problem. [Sorry for the late comment, I am just now catching up to a couple of months' worth of agda-stdlib activity!]